client/lib/create-cached-selector/index.js

```diff
@@ -0,0 +1,102 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { isObject, forEach, some, isFunction } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import warn from 'lib/warn';
+
+/*
+ * A map that is Weak with objects but Strong with primitives
+ */
+export class MixedMap {
+	weakMap = new WeakMap();
+	map = new Map();
+
+	constructor( init ) {
+		forEach( init, ( [ key, val ] ) => {
+			this.mapForKey( key ).set( key, val );
+		} );
```

Comment:
did you see this on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)

```
var kvArray = [['key1', 'value1'], ['key2', 'value2']];

// Use the regular Map constructor to transform a 2D key-value Array into a map
var myMap = new Map(kvArray);
```

Reply:
yep! which is why i needed to bake in some custom support for it.

Reply:
this constructor takes all the values that have object keys and throws them into the weakmap. the rest go into the regular map

Reply:
ah right. good good